Firmware/Chameleon-Mini/Application/MifareClassic.c

```diff
@@ -11,39 +11,40 @@
 #include "../Codec/ISO14443-2A.h"
 #include "../Memory.h"
 #include "Crypto1.h"
-#include "../Random.h"
-#include "../LED.h"
 
 #define MFCLASSIC_1K_ATQA_VALUE     0x0004
 #define MFPlus_ATQA_VALUE           0x0044
 #define MFCLASSIC_4K_ATQA_VALUE     0x0002
 #define MFCLASSIC_1K_SAK_CL1_VALUE  0x08
 #define MFCLASSIC_4K_SAK_CL1_VALUE  0x18
-#define SAK_CL1_VALUE				ISO14443A_SAK_INCOMPLETE
-#define SAK_CL2_VALUE				ISO14443A_SAK_COMPLETE_NOT_COMPLIANT
+#define SAK_CL1_VALUE               ISO14443A_SAK_INCOMPLETE
+#define SAK_CL2_VALUE               ISO14443A_SAK_COMPLETE_NOT_COMPLIANT
 
+#define MEM_S0B0_ADDRESS            0x00
+#define MEM_INVALID_ADDRESS         0xFF
 #define MEM_UID_CL1_ADDRESS         0x00
 #define MEM_UID_CL1_SIZE            4
 #define MEM_UID_BCC1_ADDRESS        0x04
-#define MEM_UID_CL2_ADDRESS        	0x03
+#define MEM_UID_CL2_ADDRESS         0x03
 #define MEM_UID_CL2_SIZE            4
+#define MEM_UID7_SIZE               7         /* Bits */
 #define MEM_KEY_A_OFFSET            48        /* Bytes */
 #define MEM_KEY_B_OFFSET            58        /* Bytes */
-#define MEM_KEY_BIGSECTOR_OFFSET	192
+#define MEM_KEY_BIGSECTOR_OFFSET    192
 #define MEM_KEY_SIZE                6         /* Bytes */
 #define MEM_ACC_GPB_SIZE            4         /* Bytes */
 #define MEM_SECTOR_ADDR_MASK        0xFC
-#define MEM_BIGSECTOR_ADDR_MASK		0xF0
+#define MEM_BIGSECTOR_ADDR_MASK     0xF0
 #define MEM_BYTES_PER_BLOCK         16        /* Bytes */
 #define MEM_VALUE_SIZE              4         /* Bytes */
 
 #define ACK_NAK_FRAME_SIZE          4         /* Bits */
 #define ACK_VALUE                   0x0A
-#define NAK_INVALID_ARG             0x00
-#define NAK_CRC_ERROR               0x01
-#define NAK_NOT_AUTHED              0x04
-#define NAK_EEPROM_ERROR            0x05
-#define NAK_OTHER_ERROR             0x06
+/* Here are values from NXP MF1S50YYX_V1 datasheet, 9.3 */
+#define NAK_S50_BUFFEROK_OPKO       0x00
+#define NAK_S50_BUFFEROK_PARITYCRC  0x01
+#define NAK_S50_BUFFERKO_OPKO       0x02
+#define NAK_S50_BUFFERKO_PARITYCRC  0X03
```

Comment:
maybe add 0x04, 0x05, 0x06 that you removed...

Reply:
Did not find any reference document for other NAK codes than these - so I guess 3+ were just internal conventions? Changed return codes accordingly in file. If you have a source and usage scenarios somewhere, I am happy to put them back in.

Reply:
There is more Mifare datasheets than just EV1 s50.     Actually 01-09 is reserved for NAK,  and 0xA is ACK.

Reply:
From what I have found - and I am surely missing something but still:
RevG code states:
```
#define NAK_INVALID_ARG             0x00
#define NAK_CRC_ERROR               0x01
#define NAK_NOT_AUTHED              0x04
#define NAK_EEPROM_ERROR            0x05
#define NAK_OTHER_ERROR             0x06
```
MF1S50YYX_V1, which is supposed to cover all the "Classic" family, title 9.3:
```
Code (4-bit) Transfer Buffer Validity Description
Ah                                    Acknowledge (ACK)
0h           valid                    invalid operation
1h           valid                    parity or CRC error
4h           invalid                  invalid operation
5h           invalid                  parity or CRC error
```

MF1S503X, title 9.3:
```
Code (4-bit)            ACK/NAK
Ah                      Acknowledge (ACK)
0h to 9h, Bh to Fh      NAK
```

MF1S5009, title 10.3:
```
Code (4-bit)            ACK/NAK
Ah                      Acknowledge (ACK)
0h to 9h                NAK
Bh to Fh                NAK
```
I cannot find more. Sure there are others, but I cannot find any reference that would state that there are specifically used as told in RevG source code, nor evidence that standard cards actually use higher codes than 0x05.
If you have any, I would love to implement a sourced list.

I put the RevG ones back in for compatibility purpose. Anyway I had serious typos.